std/algorithm/comparison.d

```diff
@@ -892,116 +892,367 @@ nothrow pure @safe unittest
             {
-                return equal(r1.byCodeUnit, r2);
+                return false;   // they can't be equal
             }
             else
             {
-                return equal(r2.byCodeUnit, r1);
-            }
-        }
-        //Try a fast implementation when the ranges have comparable lengths
-        else static if (hasLength!Range1 && hasLength!Range2 && is(typeof(r1.length == r2.length)))
-        {
-            immutable len1 = r1.length;
-            immutable len2 = r2.length;
-            if (len1 != len2) return false; //Short circuit return
+                enum someHasIndexingAndLength = anySatisfy!(hasIndexingAndLength, Rs);
 
-            //Lengths are the same, so we need to do an actual comparison
-            //Good news is we can squeeze out a bit of performance by not checking if r2 is empty
-            for (; !r1.empty; r1.popFront(), r2.popFront())
-            {
-                if (!binaryFun!(pred)(r1.front, r2.front)) return false;
-            }
-            return true;
-        }
-        else
-        {
-            //Generic case, we have to walk both ranges making sure neither is empty
-            for (; !r1.empty; r1.popFront(), r2.popFront())
-            {
-                if (r2.empty) return false;
-                if (!binaryFun!(pred)(r1.front, r2.front)) return false;
+                static if (someHasLength)
+                {
+                    /* if any `rs` `hasLength` then `primaryRangeIndex` will index the
+                     * first of them */
+                    enum primaryRangeIndex = indexOfFirstLength;
+                    const length = rs[indexOfFirstLength].length;
+                }
+                else
+                {
+                    enum primaryRangeIndex = 0; // just pick first
+                }
+                alias primaryRange = rs[primaryRangeIndex]; // shorthand
+
+                // check lengths
+                static if (someHasLength)
+                {
+                    static foreach (idx, r; rs)
+                    {
+                        static if (idx != indexOfFirstLength)
+                        {
+                            static if (hasLength!(typeof(r)))
+                            {
+                                if (length != r.length) { return false; }
+                            }
+                        }
+                    }
+                }
+
+                // check equal contents
+                static if (someHasIndexingAndLength) // if some `Rs` needs a element offset counter
+                {
+                    size_t ei = 0; // element offset counter
+                }
+                while (true) // for each element in primary range `r`
+                {
+                    static if (hasIndexingAndLength!(typeof(primaryRange)))
+                    {
+                        if (ei == primaryRange.length) { break; }
```

Comment:
No need for curly braces if you already your the one-line "shorthand"

Reply:
Done.